core/src/main/java/com/amplifyframework/core/DeveloperMenuManager.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.core;
+
+import android.content.Context;
+import android.content.Intent;
+
+/**
+ * Manage the display and shake detection behavior for the
+ * developer menu.
+ */
+public final class DeveloperMenuManager implements ShakeDetector.Listener {
+    // An instance of DeveloperMenuManager.
+    private static DeveloperMenuManager sInstance = new DeveloperMenuManager();
+    // Indicates whether the developer menu is visible.
+    private boolean visible;
+    // Listens for shake events.
+    private ShakeDetector shakeDetector;
+    // Action to take when the developer menu should be hidden.
+    private HideAction hideAction;
+
+    /**
+     * Constructs a new DeveloperMenuManager.
+     */
+    private DeveloperMenuManager() { }
+
+    /**
+     * Returns an instance of DeveloperMenuManager.
+     * @return a DeveloperMenuManager
+     */
+    public static DeveloperMenuManager sharedInstance() {
+        return sInstance;
+    }
+
+    /**
+     * Start listening for a shake event.
+     * @param context an Android Context
+     */
+    public void startListening(Context context) {
+        shakeDetector = new ShakeDetector(context, this);
+        shakeDetector.startDetecting();
+    }
+
+    @Override
+    public void onShakeDetected() {
+        if (visible) {
+            if (hideAction != null) {
+                hideAction.hideDeveloperMenu();
+            }
+            visible = false;
+        } else {
+            Context context = shakeDetector.getContext();
```

Comment:
It'd probably be better for DeveloperMenuManager to just keep its own `Context` property, which gets set in `startListening()`.  It isn't really the job of shakeDetector to be providing Context to others.

Reply:
I initially had a Context field in DeveloperMenuManager, but there was a warning that this caused a memory leak since you can access a static instance of DeveloperMenuManager.

Reply:
Storing `Context` within a static variable is a valid concern for memory leaks.   The same concern is also valid though for storing it in ShakeDetector, since it is still stored in a static instance (DeveloperMenuManager.sharedInstance().shakeDetector.context).    The same concern is even valid for your first iteration of this PR, where you were starting the DeveloperMenuManager from a callback inside of the `Amplify.configure` method.    In all cases, you are holding onto a reference of Context indefinitely.

It's important to understand there are two types of `Context` in Android:
1. The Application context, which lives for the lifetime of your app
2. An Activity `Context`, which lives for the lifecycle of an Activity

The important thing is not to keep references to any Activity `Context` instances, beyond the lifecycle of that Activity because that will prevent the Activity from being removed from memory when it is finished (a memory leak).  Using the Application Context is okay though, since is available for the lifecycle of your app.  The only caveat to this is that the Application Context has more limited functionality than an Activity Context - you can't use it to show a dialog view, for example. I don't think that's actually a concern for this use case though. 

Customers are *probably* passing in an Application Context to `Amplify.configure()`, since [that's what we say to do in the setup docs](https://docs.amplify.aws/lib/project-setup/create-application/q/platform/android#n4-initialize-amplify-in-the-application).   There's no guarantee though - an Activity class is actually a subclass of Context, so they might be passing in an Activity.

To account for this, if an Activity is passed in, you can get an Application context from it by calling `getApplicationContext()` on the Activity.  This is something we should probably do to ensure we ensure we aren't causing any memory leaks.    Here's [an example of how Firebase does something similar](https://github.com/firebase/firebase-android-sdk/blob/db66120ce7926205975a2dd50fad2f8798227a5b/firebase-common/src/main/java/com/google/firebase/FirebaseApp.java#L305-L311) in their equivalent of Amplify.configure.  For now, I'd suggest putting some similar logic inside the `startListening()` method.

Regarding the warning itself - where do you see that specifically?  Is it flagged by our checkstyle rules, or is it actually preventing the build from compiling?

Reply:
The warning appears when I open DeveloperMenuManager.java and it also prevents the build from succeeding. Using Context.getApplicationContext() and adding a Context parameter to the constructor makes the warning disappear (the warning is still present if I set the context field to context.getApplicationContext() inside the startListening method).